File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/MakeTest.java
package com.fincity.nocode.kirun.engine.function.system;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class MakeTest {

	private Make make = new Make();

	private Map<String, Map<String, Map<String, JsonElement>>> createOutputMap(Map<String, JsonElement> data) {
		return Map.of("step1", Map.of("output", data));
	}

	@Test
	void testSimpleObjectWithExpression() {
		JsonObject source = new JsonObject();
		source.addProperty("name", "John");
		source.addProperty("age", 30);

		JsonObject resultShape = new JsonObject();
		resultShape.add("fullName", new JsonPrimitive("{{Steps.step1.output.source.name}}"));
		resultShape.add("userAge", new JsonPrimitive("{{Steps.step1.output.source.age}}"));

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					assertTrue(result.isJsonObject());
					JsonObject obj = result.getAsJsonObject();
					assertEquals("John", obj.get("fullName").getAsString());
					assertEquals(30, obj.get("userAge").getAsInt());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testNestedObjectWithExpressions() {
		JsonObject user = new JsonObject();
		user.addProperty("firstName", "John");
		user.addProperty("lastName", "Doe");

		JsonObject address = new JsonObject();
		address.addProperty("city", "NYC");
		address.addProperty("zip", "10001");

		JsonObject source = new JsonObject();
		source.add("user", user);
		source.add("address", address);

		JsonObject personShape = new JsonObject();
		personShape.add("name", new JsonPrimitive("{{Steps.step1.output.source.user.firstName}}"));
		personShape.add("surname", new JsonPrimitive("{{Steps.step1.output.source.user.lastName}}"));

		JsonObject locationShape = new JsonObject();
		locationShape.add("cityName", new JsonPrimitive("{{Steps.step1.output.source.address.city}}"));
		locationShape.add("postalCode", new JsonPrimitive("{{Steps.step1.output.source.address.zip}}"));

		JsonObject resultShape = new JsonObject();
		resultShape.add("person", personShape);
		resultShape.add("location", locationShape);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					assertTrue(result.isJsonObject());
					JsonObject obj = result.getAsJsonObject();
					assertEquals("John", obj.getAsJsonObject("person").get("name").getAsString());
					assertEquals("Doe", obj.getAsJsonObject("person").get("surname").getAsString());
					assertEquals("NYC", obj.getAsJsonObject("location").get("cityName").getAsString());
					assertEquals("10001", obj.getAsJsonObject("location").get("postalCode").getAsString());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testArrayWithExpressions() {
		JsonArray items = new JsonArray();
		items.add("apple");
		items.add("banana");
		items.add("cherry");

		JsonObject source = new JsonObject();
		source.add("items", items);

		JsonArray fruitsShape = new JsonArray();
		fruitsShape.add(new JsonPrimitive("{{Steps.step1.output.source.items[0]}}"));
		fruitsShape.add(new JsonPrimitive("{{Steps.step1.output.source.items[1]}}"));
		fruitsShape.add(new JsonPrimitive("{{Steps.step1.output.source.items[2]}}"));

		JsonObject resultShape = new JsonObject();
		resultShape.add("fruits", fruitsShape);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					assertTrue(result.isJsonObject());
					JsonArray fruits = result.getAsJsonObject().getAsJsonArray("fruits");
					assertEquals("apple", fruits.get(0).getAsString());
					assertEquals("banana", fruits.get(1).getAsString());
					assertEquals("cherry", fruits.get(2).getAsString());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testDeeplyNestedStructureWithArrays() {
		JsonObject user1 = new JsonObject();
		user1.addProperty("id", 1);
		user1.addProperty("name", "Alice");

		JsonObject user2 = new JsonObject();
		user2.addProperty("id", 2);
		user2.addProperty("name", "Bob");

		JsonArray users = new JsonArray();
		users.add(user1);
		users.add(user2);

		JsonObject data = new JsonObject();
		data.add("users", users);

		JsonObject source = new JsonObject();
		source.add("data", data);

		JsonObject userShape1 = new JsonObject();
		userShape1.add("userId", new JsonPrimitive("{{Steps.step1.output.source.data.users[0].id}}"));
		userShape1.add("userName", new JsonPrimitive("{{Steps.step1.output.source.data.users[0].name}}"));

		JsonObject userShape2 = new JsonObject();
		userShape2.add("userId", new JsonPrimitive("{{Steps.step1.output.source.data.users[1].id}}"));
		userShape2.add("userName", new JsonPrimitive("{{Steps.step1.output.source.data.users[1].name}}"));

		JsonArray userListShape = new JsonArray();
		userListShape.add(userShape1);
		userListShape.add(userShape2);

		JsonObject level3 = new JsonObject();
		level3.add("userList", userListShape);

		JsonObject level2 = new JsonObject();
		level2.add("level3", level3);

		JsonObject level1 = new JsonObject();
		level1.add("level2", level2);

		JsonObject resultShape = new JsonObject();
		resultShape.add("level1", level1);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					JsonArray userList = result.getAsJsonObject()
							.getAsJsonObject("level1")
							.getAsJsonObject("level2")
							.getAsJsonObject("level3")
							.getAsJsonArray("userList");
					assertEquals(1, userList.get(0).getAsJsonObject().get("userId").getAsInt());
					assertEquals("Alice", userList.get(0).getAsJsonObject().get("userName").getAsString());
					assertEquals(2, userList.get(1).getAsJsonObject().get("userId").getAsInt());
					assertEquals("Bob", userList.get(1).getAsJsonObject().get("userName").getAsString());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testMixedStaticAndDynamicValues() {
		JsonObject source = new JsonObject();
		source.addProperty("dynamicValue", "from source");

		JsonObject nestedShape = new JsonObject();
		nestedShape.addProperty("staticNum", 42);
		nestedShape.add("dynamicNum", new JsonPrimitive("{{Steps.step1.output.source.dynamicValue}}"));

		JsonArray arrayShape = new JsonArray();
		arrayShape.add("static");
		arrayShape.add(new JsonPrimitive("{{Steps.step1.output.source.dynamicValue}}"));
		arrayShape.add(123);

		JsonObject resultShape = new JsonObject();
		resultShape.addProperty("static", "static string");
		resultShape.add("dynamic", new JsonPrimitive("{{Steps.step1.output.source.dynamicValue}}"));
		resultShape.add("nested", nestedShape);
		resultShape.add("array", arrayShape);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					JsonObject obj = result.getAsJsonObject();
					assertEquals("static string", obj.get("static").getAsString());
					assertEquals("from source", obj.get("dynamic").getAsString());
					assertEquals(42, obj.getAsJsonObject("nested").get("staticNum").getAsInt());
					assertEquals("from source", obj.getAsJsonObject("nested").get("dynamicNum").getAsString());
					JsonArray arr = obj.getAsJsonArray("array");
					assertEquals("static", arr.get(0).getAsString());
					assertEquals("from source", arr.get(1).getAsString());
					assertEquals(123, arr.get(2).getAsInt());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testNullHandling() {
		JsonObject nestedShape = new JsonObject();
		nestedShape.add("inner", JsonNull.INSTANCE);

		JsonObject resultShape = new JsonObject();
		resultShape.add("nullValue", JsonNull.INSTANCE);
		resultShape.add("nested", nestedShape);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(Map.of());

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					JsonObject obj = result.getAsJsonObject();
					assertTrue(obj.get("nullValue").isJsonNull());
					assertTrue(obj.getAsJsonObject("nested").get("inner").isJsonNull());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testArrayOfArrays() {
		JsonArray row1 = new JsonArray();
		row1.add(1);
		row1.add(2);

		JsonArray row2 = new JsonArray();
		row2.add(3);
		row2.add(4);

		JsonArray matrix = new JsonArray();
		matrix.add(row1);
		matrix.add(row2);

		JsonObject source = new JsonObject();
		source.add("matrix", matrix);

		JsonArray gridRow1 = new JsonArray();
		gridRow1.add(new JsonPrimitive("{{Steps.step1.output.source.matrix[0][0]}}"));
		gridRow1.add(new JsonPrimitive("{{Steps.step1.output.source.matrix[0][1]}}"));

		JsonArray gridRow2 = new JsonArray();
		gridRow2.add(new JsonPrimitive("{{Steps.step1.output.source.matrix[1][0]}}"));
		gridRow2.add(new JsonPrimitive("{{Steps.step1.output.source.matrix[1][1]}}"));

		JsonArray gridShape = new JsonArray();
		gridShape.add(gridRow1);
		gridShape.add(gridRow2);

		JsonObject resultShape = new JsonObject();
		resultShape.add("grid", gridShape);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertNotNull(result);
					JsonArray grid = result.getAsJsonObject().getAsJsonArray("grid");
					assertEquals(1, grid.get(0).getAsJsonArray().get(0).getAsInt());
					assertEquals(2, grid.get(0).getAsJsonArray().get(1).getAsInt());
					assertEquals(3, grid.get(1).getAsJsonArray().get(0).getAsInt());
					assertEquals(4, grid.get(1).getAsJsonArray().get(1).getAsInt());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testPrimitiveResultShape() {
		JsonObject source = new JsonObject();
		source.addProperty("value", "hello");

		JsonPrimitive resultShape = new JsonPrimitive("{{Steps.step1.output.source.value}}");

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertEquals("hello", result.getAsString());
					return true;
				})
				.verifyComplete();
	}

	@Test
	void testArrayAsRootResultShape() {
		JsonObject source = new JsonObject();
		source.addProperty("a", 1);
		source.addProperty("b", 2);
		source.addProperty("c", 3);

		JsonArray resultShape = new JsonArray();
		resultShape.add(new JsonPrimitive("{{Steps.step1.output.source.a}}"));
		resultShape.add(new JsonPrimitive("{{Steps.step1.output.source.b}}"));
		resultShape.add(new JsonPrimitive("{{Steps.step1.output.source.c}}"));

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("resultShape", resultShape))
				.setContext(Map.of())
				.setSteps(createOutputMap(Map.of("source", source)));

		StepVerifier.create(make.execute(fep)
				.map(output -> output.allResults().getFirst().getResult().get("value")))
				.expectNextMatches(result -> {
					assertTrue(result.isJsonArray());
					JsonArray arr = result.getAsJsonArray();
					assertEquals(1, arr.get(0).getAsInt());
					assertEquals(2, arr.get(1).getAsInt());
					assertEquals(3, arr.get(2).getAsInt());
					return true;
				})
				.verifyComplete();
	}
}

File: kirun-js/__tests__/engine/function/system/MakeTest.ts
import {
    FunctionExecutionParameters,
    KIRunFunctionRepository,
    KIRunSchemaRepository,
} from '../../../../src';
import { Make } from '../../../../src/engine/function/system/Make';

const make: Make = new Make();

function createOutputMap(data: Record<string, any>): Map<string, Map<string, Map<string, any>>> {
    return new Map([
        [
            'step1',
            new Map([
                [
                    'output',
                    new Map(Object.entries(data)),
                ],
            ]),
        ],
    ]);
}

test('test simple object with expression', async () => {
    const source = { name: 'John', age: 30 };

    const resultShape = {
        fullName: '{{Steps.step1.output.source.name}}',
        userAge: '{{Steps.step1.output.source.age}}',
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        fullName: 'John',
        userAge: 30,
    });
});

test('test nested object with expressions', async () => {
    const source = {
        user: { firstName: 'John', lastName: 'Doe' },
        address: { city: 'NYC', zip: '10001' },
    };

    const resultShape = {
        person: {
            name: '{{Steps.step1.output.source.user.firstName}}',
            surname: '{{Steps.step1.output.source.user.lastName}}',
        },
        location: {
            cityName: '{{Steps.step1.output.source.address.city}}',
            postalCode: '{{Steps.step1.output.source.address.zip}}',
        },
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        person: {
            name: 'John',
            surname: 'Doe',
        },
        location: {
            cityName: 'NYC',
            postalCode: '10001',
        },
    });
});

test('test array with expressions', async () => {
    const source = {
        items: ['apple', 'banana', 'cherry'],
    };

    const resultShape = {
        fruits: [
            '{{Steps.step1.output.source.items[0]}}',
            '{{Steps.step1.output.source.items[1]}}',
            '{{Steps.step1.output.source.items[2]}}',
        ],
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        fruits: ['apple', 'banana', 'cherry'],
    });
});

test('test deeply nested structure with arrays', async () => {
    const source = {
        data: {
            users: [
                { id: 1, name: 'Alice' },
                { id: 2, name: 'Bob' },
            ],
        },
    };

    const resultShape = {
        level1: {
            level2: {
                level3: {
                    userList: [
                        {
                            userId: '{{Steps.step1.output.source.data.users[0].id}}',
                            userName: '{{Steps.step1.output.source.data.users[0].name}}',
                        },
                        {
                            userId: '{{Steps.step1.output.source.data.users[1].id}}',
                            userName: '{{Steps.step1.output.source.data.users[1].name}}',
                        },
                    ],
                },
            },
        },
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        level1: {
            level2: {
                level3: {
                    userList: [
                        { userId: 1, userName: 'Alice' },
                        { userId: 2, userName: 'Bob' },
                    ],
                },
            },
        },
    });
});

test('test mixed static and dynamic values', async () => {
    const source = { dynamicValue: 'from source' };

    const resultShape = {
        static: 'static string',
        dynamic: '{{Steps.step1.output.source.dynamicValue}}',
        nested: {
            staticNum: 42,
            dynamicNum: '{{Steps.step1.output.source.dynamicValue}}',
        },
        array: ['static', '{{Steps.step1.output.source.dynamicValue}}', 123],
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        static: 'static string',
        dynamic: 'from source',
        nested: {
            staticNum: 42,
            dynamicNum: 'from source',
        },
        array: ['static', 'from source', 123],
    });
});

test('test null and undefined handling', async () => {
    const resultShape = {
        nullValue: null,
        nested: {
            inner: null,
        },
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(new Map());

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        nullValue: null,
        nested: {
            inner: null,
        },
    });
});

test('test array of arrays', async () => {
    const source = {
        matrix: [
            [1, 2],
            [3, 4],
        ],
    };

    const resultShape = {
        grid: [
            ['{{Steps.step1.output.source.matrix[0][0]}}', '{{Steps.step1.output.source.matrix[0][1]}}'],
            ['{{Steps.step1.output.source.matrix[1][0]}}', '{{Steps.step1.output.source.matrix[1][1]}}'],
        ],
    };

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual({
        grid: [
            [1, 2],
            [3, 4],
        ],
    });
});

test('test primitive result shape', async () => {
    const source = { value: 'hello' };

    const resultShape = '{{Steps.step1.output.source.value}}';

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toBe('hello');
});

test('test array as root result shape', async () => {
    const source = { a: 1, b: 2, c: 3 };

    const resultShape = [
        '{{Steps.step1.output.source.a}}',
        '{{Steps.step1.output.source.b}}',
        '{{Steps.step1.output.source.c}}',
    ];

    const fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    fep.setArguments(new Map<string, any>([['resultShape', resultShape]]));
    fep.setContext(new Map());
    fep.setSteps(createOutputMap({ source }));

    const result = await make.execute(fep);
    const value = result.allResults()[0]?.getResult()?.get('value');

    expect(value).toStrictEqual([1, 2, 3]);
});
